Libraries/Opc.Ua.Gds.Client.Common/GlobalDiscoveryServerClient.cs

```diff
@@ -1471,24 +1422,97 @@ public async Task<TrustListDataType> ReadTrustListAsync(NodeId trustListId, Canc
             }
             finally
             {
-                if (IsConnected)
-                {
-                    await Session.CallAsync(trustListId, Ua.MethodIds.FileType_Close, ct, fileHandle).ConfigureAwait(false);
-                }
+                await session.CallAsync(trustListId, Ua.MethodIds.FileType_Close, ct, fileHandle).ConfigureAwait(false);
             }
 
             ostrm.Position = 0;
 
             var trustList = new TrustListDataType();
-            using (var decoder = new BinaryDecoder(ostrm, Session.MessageContext))
+            using (var decoder = new BinaryDecoder(ostrm, session.MessageContext))
             {
                 trustList.Decode(decoder);
             }
             return trustList;
         }
 
-        private ConfiguredEndpoint m_endpoint;
+        /// <summary>
+        /// Connect the session
+        /// </summary>
+        private async Task ConnectInternalAsync(
+            ConfiguredEndpoint endpoint,
+            bool updateBeforeConnect,
+            CancellationToken ct)
+        {
+            lock (m_lock)
```

Comment:
Using thread locking primitives inside async code is very dangerous. - The thread that runs the current task may be the thread that is locked.